code/object/object.cpp

```diff
@@ -1593,15 +1560,27 @@ void obj_move_all(float frametime)
 				// physics
 				obj_move_call_physics(objp, frametime);
 			}
-		} else {
-			// make sure velocity is always 0 for immobile things!
+		}
+
+		// if the object isn't supposed to move, roll back the movement
```

Comment:
I would be slightly more specific here, these *should* already be the case. But are here as just as a last-line-of-defense to ensure collision or anything else in the future doesn't accidentally let velocity or position changes 'leak' out. We don't want to accidentally imply that movement, despite these flags, is expected and therefore needs to be rolled back.

Reply:
It's not really a last line of defense; in many cases it's a first line of defense.  Without these lines, a ship flying waypoints would continue to move, and a ship blasted by a shockwave would change position.  (There might be other cases too.)  So yes, movement is expected here.

Reply:
A ship wouldn't be able to turn because of the skip in `ai_turn_towards_vector` and wouldn't be able to move under it's own power because of the skip immediately before this section, although a shockwave would still go through. This section is quite new, and so for a long immobile went without this and still sort-of worked. There *are* cases for which this is necessary, but I would consider them unusual and undesirable, I don't want other coders seeing this and thinking that because this exists, immobility doesn't need to be handled elsewhere, because it'll just get rolled back in the end.

Reply:
The section isn't new; compare the previous version of object.cpp.  Previously, `obj_move_call_physics` was skipped and velocity (both linear and rotational) was zeroed.  The new code is similar, but treats linear and rotational velocity individually, and only skips physics if *both* flags are in effect.  I considered modifying `obj_move_call_physics` to check the flags individually, but the code just got too tangled.

To the original point, I can enhance the code comment.  What about something like this:
```
// If the object isn't supposed to move, roll back any movement that occurred.  Most of the movement should already have been skipped, but this ensures complete immobility.
```

Reply:
I'd be fine with that.

Reply:
Done.